### PR TITLE
[consensus] send to distant peers first on broadcast

### PR DESCRIPTION
**Pull request by ibalajiarun**
## Description

It seems beneficial to prioritize sending messages to distant peers first during a broadcast.

This PR introduces a method to `PeersAndMetadata` to sort peers by latency. 

- NetworkClient uses this sorting for `send_to_many` calls. 
- RBNetworkSender trait exposes a method to sort peers by latency for Reliable Broadcast RPCs.
- Removes cloning for `PeerMonitoringMetadata` in `PeerMetadata`, because the clone is expensive about 9 microseconds per clone with 100 validators.

## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Breaking change
- [x] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by zekun000 — COMMENTED** [review]
it looks a bit weird living outside of the network interface, how about we include this as part of this one? https://github.com/aptos-labs/aptos-core/blob/d9666758ae978e03ff4fbf7288a5432741d475b6/consensus/src/network_interface.rs#L123

**Comment by ibalajiarun** [comment]
> it looks a bit weird living outside of the network interface, how about we include this as part of this one?

If we put this under consensus, then we cannot use this in other places like dkg, so I put it under network code.

**Comment by zekun000** [comment]
we already have the peers_and_metadata in the network client right? https://github.com/aptos-labs/aptos-core/blob/601b59774696438f8329b14c378b19ebcf7cd181/network/framework/src/application/interface.rs#L73

just need to expose a function for the trait?

**Comment by ibalajiarun** [comment]
> we already have the peers_and_metadata in the network client right?

okay, I think i have simplified it now. I didn't realize networkclient has peers and metadata.

File: network/framework/src/application/metadata.rs [hunk 1]
```diff
@@ -87,6 +88,14 @@ impl PeerMetadata {
 
     /// Returns a copy of the peer monitoring metadata
     pub fn get_peer_monitoring_metadata(&self) -> PeerMonitoringMetadata {
+        let _timer = counters::MONITOR
+            .with_label_values(&["metadata_clone"])
+            .start_timer();
         self.peer_monitoring_metadata.clone()
     }
+
+    /// Returns a copy of the peer monitoring metadata
```
Comment (zekun000) [inline network/framework/src/application/metadata.rs:97]: s/copy/ref

File: network/framework/src/application/metadata.rs [hunk 2]
```diff
@@ -87,6 +88,14 @@ impl PeerMetadata {
 
     /// Returns a copy of the peer monitoring metadata
     pub fn get_peer_monitoring_metadata(&self) -> PeerMonitoringMetadata {
```
Comment (zekun000) [inline network/framework/src/application/metadata.rs:90]: can we remove this expensive function? and even remove the Clone from the struct?

File: network/framework/src/application/storage.rs [hunk 3]
```diff
@@ -38,7 +39,7 @@ pub struct PeersAndMetadata {
     // but infrequent writes. The cache is updated on all underlying updates.
     //
     // TODO: should we remove this when generational versioning is supported?
-    cached_peers_and_metadata: Arc<ArcSwap<HashMap<NetworkId, HashMap<PeerId, PeerMetadata>>>>,
+    cached_peers_and_metadata: Arc<RwLock<HashMap<NetworkId, HashMap<PeerId, PeerMetadata>>>>,
```
Comment (zekun000) [inline network/framework/src/application/storage.rs:42]: we can revert this change?

File: network/framework/src/application/metadata.rs (hunk 2 above)
Comment (ibalajiarun) [reply network/framework/src/application/metadata.rs]: I did, but I did not remove clone because it's appears necessary in some places.

**Review by zekun000 — APPROVED** [review]
thanks for cleaning up the clone!

**Review by sitalkedia — APPROVED** [review]
Did you observe any improvement in e2e latency with this change? I had tried similar changes in past and didn't see any improvement.
